File: freeswitch_client.py
"""
Freeswitch ESL????????Freeswitch?????
"""
import socket
import threading
import logging
from typing import Callable, Optional, Dict
from config import Config

logger = logging.getLogger(__name__)


class FreeswitchClient:
    """Freeswitch ESL???"""
    
    def __init__(self, host: str = None, port: int = None, password: str = None):
        self.host = host or Config.FS_HOST
        self.port = port or Config.FS_PORT
        self.password = password or Config.FS_PASSWORD
        self.socket: Optional[socket.socket] = None
        self.connected = False
        self.event_handlers: Dict[str, Callable] = {}
        self.running = False
        self.thread: Optional[threading.Thread] = None
        
    def connect(self) -> bool:
        """???Freeswitch"""
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.connect((self.host, self.port))
            
            # ??????
            welcome = self._read_response()
            logger.info(f"Freeswitch????: {welcome}")
            
            # ??
            auth_response = self._send_command(f"auth {self.password}")
            if "OK" in auth_response:
                self.connected = True
                logger.info("Freeswitch????")
                return True
            else:
                logger.error(f"Freeswitch????: {auth_response}")
                return False
                
        except Exception as e:
            logger.error(f"??Freeswitch??: {e}")
            return False
    
    def _send_command(self, command: str) -> str:
        """?????Freeswitch"""
        if not self.socket:
            return ""
        
        try:
            command = command + "\n\n"
            self.socket.send(command.encode('utf-8'))
            return self._read_response()
        except Exception as e:
            logger.error(f"??????: {e}")
            return ""
    
    def _read_response(self) -> str:
        """??Freeswitch??"""
        if not self.socket:
            return ""
        
        try:
            response = ""
            while True:
                data = self.socket.recv(4096).decode('utf-8', errors='ignore')
                if not data:
                    break
                response += data
                if "\n\n" in response or "\r\n\r\n" in response:
                    break
            return response
        except Exception as e:
            logger.error(f"??????: {e}")
            return ""
    
    def subscribe_events(self, events: list = None):
        """????"""
        if events is None:
            events = ["CHANNEL_CREATE", "CHANNEL_ANSWER", "CHANNEL_BRIDGE", 
                     "CHANNEL_HANGUP", "CUSTOM", "sofia::register", 
                     "sofia::unregister", "sofia::gateway-add", 
                     "sofia::gateway-state"]
        
        event_string = " ".join(events)
        response = self._send_command(f"events plain {event_string}")
        logger.info(f"????: {response}")
    
    def register_event_handler(self, event_type: str, handler: Callable):
        """???????"""
        self.event_handlers[event_type] = handler
    
    def start_listening(self):
        """??????"""
        if not self.connected:
            logger.error("????Freeswitch")
            return
        
        self.running = True
        self.thread = threading.Thread(target=self._event_loop, daemon=True)
        self.thread.start()
        logger.info("????Freeswitch??")
    
    def _event_loop(self):
        """????"""
        while self.running and self.connected:
            try:
                event_data = self._read_response()
                if event_data:
                    self._process_event(event_data)
            except Exception as e:
                logger.error(f"??????: {e}")
                if self.running:
                    # ????
                    self.connected = False
                    self.connect()
    
    def _process_event(self, event_data: str):
        """????"""
        event_dict = {}
        for line in event_data.split('\n'):
            if ':' in line:
                key, value = line.split(':', 1)
                event_dict[key.strip()] = value.strip()
        
        event_name = event_dict.get('Event-Name', '')
        logger.debug(f"????: {event_name}")
        
        # ????????
        if event_name in self.event_handlers:
            try:
                self.event_handlers[event_name](event_dict)
            except Exception as e:
                logger.error(f"???? {event_name} ??: {e}")
    
    def execute_command(self, command: str, args: str = "") -> str:
        """??Freeswitch??"""
        full_command = f"api {command} {args}"
        return self._send_command(full_command)
    
    def uuid_record(self, uuid: str, file_path: str, channels: str = "both"):
        """????"""
        command = f"uuid_record {uuid} start {file_path} {channels}"
        return self.execute_command("uuid_record", f"{uuid} start {file_path} {channels}")
    
    def uuid_bridge(self, uuid1: str, uuid2: str):
        """????UUID"""
        return self.execute_command("uuid_bridge", f"{uuid1} {uuid2}")
    
    def disconnect(self):
        """????"""
        self.running = False
        if self.socket:
            try:
                self.socket.close()
            except:
                pass
        self.connected = False
        logger.info("???Freeswitch??")

File: src/main/java/com/ivr/system/esl/FreeSwitchClient.java
package com.ivr.system.esl;

import com.ivr.system.config.FreeSwitchConfig;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.*;
import java.net.Socket;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.*;

/**
 * FreeSWITCH ESL客户端
 * 实现与FreeSWITCH的Event Socket Library通信
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class FreeSwitchClient {

    private final FreeSwitchConfig config;

    private Socket socket;
    private BufferedReader reader;
    private PrintWriter writer;
    private volatile boolean connected = false;

    private final ExecutorService executorService = Executors.newCachedThreadPool();
    private final Map<String, CompletableFuture<EslResponse>> pendingCommands = new ConcurrentHashMap<>();
    private final Map<String, EslEventListener> eventListeners = new ConcurrentHashMap<>();

    @PostConstruct
    public void init() {
        if (config.isAutoConnect()) {
            connect();
        }
    }

    @PreDestroy
    public void destroy() {
        disconnect();
        executorService.shutdown();
    }

    /**
     * 连接到FreeSWITCH
     */
    public synchronized boolean connect() {
        if (connected) {
            return true;
        }

        try {
            log.info("Connecting to FreeSWITCH at {}:{}", config.getHost(), config.getPort());

            socket = new Socket(config.getHost(), config.getPort());
            socket.setSoTimeout(config.getTimeout());
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            writer = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()), true);

            EslResponse authResponse = readResponse();
            if (authResponse == null || !authResponse.isSuccess()) {
                log.error("Failed to receive auth challenge");
                return false;
            }

            sendCommand("auth " + config.getPassword());
            EslResponse loginResponse = readResponse();

            if (loginResponse != null && loginResponse.isSuccess()) {
                connected = true;
                log.info("Successfully connected to FreeSWITCH");

                subscribeEvents("all");

                startEventListener();
                return true;
            } else {
                log.error("Authentication failed");
                disconnect();
                return false;
            }
        } catch (Exception e) {
            log.error("Failed to connect to FreeSWITCH: {}", e.getMessage());
            disconnect();
            return false;
        }
    }

    /**
     * 断开连接
     */
    public synchronized void disconnect() {
        connected = false;
        try {
            if (writer != null) writer.close();
            if (reader != null) reader.close();
            if (socket != null) socket.close();
        } catch (IOException e) {
            log.error("Error disconnecting: {}", e.getMessage());
        }
        log.info("Disconnected from FreeSWITCH");
    }

    /**
     * 发送命令
     */
    private void sendCommand(String command) {
        if (writer != null) {
            writer.println(command);
            writer.println();
            writer.flush();
        }
    }

    /**
     * 读取响应
     */
    private EslResponse readResponse() throws IOException {
        Map<String, String> headers = new HashMap<>();
        String line;

        while ((line = reader.readLine()) != null && !line.isEmpty()) {
            int colonPos = line.indexOf(':');
            if (colonPos > 0) {
                String key = line.substring(0, colonPos).trim();
                String value = line.substring(colonPos + 1).trim();
                headers.put(key, value);
            }
        }

        String body = null;
        if (headers.containsKey("Content-Length")) {
            int length = Integer.parseInt(headers.get("Content-Length"));
            char[] buffer = new char[length];
            int read = reader.read(buffer, 0, length);
            if (read > 0) {
                body = new String(buffer, 0, read);
            }
        }

        return new EslResponse(headers, body);
    }

    /**
     * 订阅事件
     */
    public void subscribeEvents(String events) {
        sendCommand("event plain " + events);
    }

    /**
     * 启动事件监听器
     */
    private void startEventListener() {
        executorService.submit(() -> {
            while (connected) {
                try {
                    EslResponse response = readResponse();
                    if (response != null) {
                        handleEvent(response);
                    }
                } catch (IOException e) {
                    if (connected) {
                        log.error("Error reading event: {}", e.getMessage());
                    }
                }
            }
        });
    }

    /**
     * 处理事件
     */
    private void handleEvent(EslResponse response) {
        String eventName = response.getHeader("Event-Name");
        String uuid = response.getHeader("Unique-ID");

        if (eventName != null) {
            log.debug("Received event: {} for UUID: {}", eventName, uuid);

            eventListeners.forEach((key, listener) -> {
                try {
                    listener.onEvent(eventName, uuid, response);
                } catch (Exception e) {
                    log.error("Error in event listener: {}", e.getMessage());
                }
            });
        }
    }

    /**
     * 注册事件监听器
     */
    public String addEventListener(EslEventListener listener) {
        String listenerId = UUID.randomUUID().toString();
        eventListeners.put(listenerId, listener);
        return listenerId;
    }

    /**
     * 移除事件监听器
     */
    public void removeEventListener(String listenerId) {
        eventListeners.remove(listenerId);
    }

    /**
     * 执行API命令
     */
    public EslResponse api(String command) {
        return execute("api " + command);
    }

    /**
     * 执行bgapi命令
     */
    public String bgapi(String command) {
        EslResponse response = execute("bgapi " + command);
        if (response != null && response.getBody() != null) {
            return response.getBody().replace("Job-UUID: ", "").trim();
        }
        return null;
    }

    /**
     * 执行命令
     */
    private synchronized EslResponse execute(String command) {
        if (!connected) {
            log.warn("Not connected to FreeSWITCH");
            return null;
        }

        try {
            sendCommand(command);
            return readResponse();
        } catch (IOException e) {
            log.error("Error executing command: {}", e.getMessage());
            return null;
        }
    }

    /**
     * 播放音频文件
     */
    public boolean playback(String uuid, String file) {
        log.info("Playing file [{}] for UUID [{}]", file, uuid);
        EslResponse response = api("uuid_broadcast " + uuid + " " + file + " aleg");
        return response != null && response.isSuccess();
    }

    /**
     * 使用TTS播放文本
     */
    public boolean speak(String uuid, String text) {
        log.info("Speaking text for UUID [{}]: {}", uuid, text);
        String ttsEngine = config.getTtsEngine();
        String ttsVoice = config.getTtsVoice();
        EslResponse response = api("uuid_broadcast " + uuid +
                " speak:" + ttsEngine + "|" + ttsVoice + "|" + text + " aleg");
        return response != null && response.isSuccess();
    }

    /**
     * 开始DTMF收集
     */
    public boolean startDtmfCollection(String uuid, int maxDigits, long timeoutMs) {
        log.info("Starting DTMF collection for UUID [{}], max digits: {}, timeout: {}ms",
                uuid, maxDigits, timeoutMs);

        api("uuid_setvar " + uuid + " max_digits " + maxDigits);
        api("uuid_setvar " + uuid + " digit_timeout " + (timeoutMs / 1000));

        return true;
    }

    /**
     * 停止DTMF收集
     */
    public boolean stopDtmfCollection(String uuid) {
        log.info("Stopping DTMF collection for UUID [{}]", uuid);
        return true;
    }

    /**
     * 转接到坐席
     */
    public boolean transferToAgent(String uuid, String agentGroup, String skillGroup) {
        log.info("Transferring UUID [{}] to agent group [{}], skill [{}]",
                uuid, agentGroup, skillGroup);

        String dialString = "user/" + agentGroup;
        if (skillGroup != null && !skillGroup.isEmpty()) {
            dialString = "loopback/queue_" + skillGroup;
        }

        EslResponse response = api("uuid_transfer " + uuid + " " + dialString);
        return response != null && response.isSuccess();
    }

    /**
     * 转接到外部号码
     */
    public boolean transferToExternal(String uuid, String destination, String gateway) {
        log.info("Transferring UUID [{}] to external [{}] via gateway [{}]",
                uuid, destination, gateway);

        String dialString = "sofia/gateway/" + gateway + "/" + destination;
        EslResponse response = api("uuid_transfer " + uuid + " " + dialString);
        return response != null && response.isSuccess();
    }

    /**
     * 挂断通话
     */
    public boolean hangup(String uuid, String cause) {
        log.info("Hanging up UUID [{}] with cause [{}]", uuid, cause);
        EslResponse response = api("uuid_kill " + uuid + " " + cause);
        return response != null && response.isSuccess();
    }

    /**
     * 开始录音
     */
    public boolean startRecording(String uuid, String filePath, int maxDuration, int silenceThreshold) {
        log.info("Starting recording for UUID [{}] to file [{}]", uuid, filePath);
        EslResponse response = api("uuid_record " + uuid + " start " + filePath +
                " " + maxDuration + " " + silenceThreshold);
        return response != null && response.isSuccess();
    }

    /**
     * 停止录音
     */
    public boolean stopRecording(String uuid) {
        log.info("Stopping recording for UUID [{}]", uuid);
        EslResponse response = api("uuid_record " + uuid + " stop");
        return response != null && response.isSuccess();
    }

    /**
     * 进入队列
     */
    public boolean enterQueue(String uuid, String queueName, String holdMusic) {
        log.info("Entering queue [{}] for UUID [{}]", queueName, uuid);

        if (holdMusic != null && !holdMusic.isEmpty()) {
            api("uuid_setvar " + uuid + " hold_music " + holdMusic);
        }

        EslResponse response = api("uuid_transfer " + uuid + " callcenter:" + queueName);
        return response != null && response.isSuccess();
    }

    /**
     * 开始ASR识别
     */
    public boolean startAsrRecognition(String uuid, String engine, String grammar, long timeout) {
        log.info("Starting ASR for UUID [{}] with engine [{}]", uuid, engine);

        api("uuid_setvar " + uuid + " asr_engine " + engine);
        if (grammar != null) {
            api("uuid_setvar " + uuid + " asr_grammar " + grammar);
        }
        api("uuid_setvar " + uuid + " asr_timeout " + (timeout / 1000));

        EslResponse response = api("uuid_broadcast " + uuid + " detect_speech aleg");
        return response != null && response.isSuccess();
    }

    /**
     * 发起呼出
     */
    public String originate(String destination, String callerIdNumber, String callerIdName,
                            String gateway, Map<String, String> variables) {
        log.info("Originating call to [{}] from [{}]", destination, callerIdNumber);

        StringBuilder dialString = new StringBuilder();
        dialString.append("{");

        if (callerIdNumber != null) {
            dialString.append("origination_caller_id_number=").append(callerIdNumber).append(",");
        }
        if (callerIdName != null) {
            dialString.append("origination_caller_id_name=").append(callerIdName).append(",");
        }

        if (variables != null) {
            variables.forEach((k, v) -> dialString.append(k).append("=").append(v).append(","));
        }

        if (dialString.charAt(dialString.length() - 1) == ',') {
            dialString.setLength(dialString.length() - 1);
        }

        dialString.append("}");
        dialString.append("sofia/gateway/").append(gateway).append("/").append(destination);

        String jobUuid = bgapi("originate " + dialString + " &park()");
        log.info("Originate job UUID: {}", jobUuid);
        return jobUuid;
    }

    /**
     * 桥接两个通话
     */
    public boolean bridge(String uuid1, String uuid2) {
        log.info("Bridging UUID [{}] with UUID [{}]", uuid1, uuid2);
        EslResponse response = api("uuid_bridge " + uuid1 + " " + uuid2);
        return response != null && response.isSuccess();
    }

    /**
     * 设置通道变量
     */
    public boolean setVariable(String uuid, String variable, String value) {
        EslResponse response = api("uuid_setvar " + uuid + " " + variable + " " + value);
        return response != null && response.isSuccess();
    }

    /**
     * 获取通道变量
     */
    public String getVariable(String uuid, String variable) {
        EslResponse response = api("uuid_getvar " + uuid + " " + variable);
        if (response != null && response.getBody() != null) {
            return response.getBody().trim();
        }
        return null;
    }

    /**
     * 检查连接状态
     */
    public boolean isConnected() {
        return connected;
    }

    /**
     * 事件监听器接口
     */
    public interface EslEventListener {
        void onEvent(String eventName, String uuid, EslResponse response);
    }
}
